scalding-beam improvements on resulting DAG

tlazaro opened the pull request:
- Implements missing Merge and Fork
- Improvement to have names and better grouping of some operations so these show as fewer boxes on the Dataflow UI. These boxes can be expanded if needed to show the inner steps.
- Caching of runs of BeamOp now means PCollection used by multiple steps don't recompute.

tlazaro committed:
Beam backend PTransform names. Merge and Fork impl.

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala:
```diff
@@ -176,6 +180,54 @@ object BeamOp extends Serializable {
+        widenPCollection(rightPCollection): PCollection[(K, _)]
+      )
+
+      tuple.apply(HashJoinTransform(keyCoder, joiner))
```
we don't need the explicit `apply` do we ? `tuple(HashJoinTransform(keyCoder, joiner))` would work, no?

tlazaro replied:
Yes that is true but for all the `apply` from Beam's Pipeline, PTransforms and similar we write it explicitly to signal it's Beam API calls and that's what it would look like called from Java or from Beam examples.

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala:
```diff
@@ -176,6 +180,54 @@ object BeamOp extends Serializable {
+    }
+  }
+
+  final case class MergedBeamOp[A](operations: Seq[BeamOp[A]]) extends BeamOp[A] {
```
do we want to add `require(operations.size > 1, s"expected at least two items in Merged: $operations")` just to guard against a refactoring creating suboptimal plans here?

tlazaro replied:
Sounds reasonable, or we could make it `first`, `second`, `tail` ?

johnynek reviewed (commented):
this looks good.

a few comments.

Also, I think toIterableExecution and forceToDiskExecution aren't implemented yet. Is that right?

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamBackend.scala:
```diff
@@ -115,6 +115,14 @@ object BeamPlanner {
           }
           go(cg)
+        case (Fork(input), rec) =>
+          rec(input)
```
btw: if you could link to any documentation here that says that beam can handle fanouts well or under what circumstances you need to annotate them in any way, that would be useful.

A fork node is added by the planner if a node has more than one downstream consumer.

tlazaro replied:
I assumed it would but let me dig into that.

tlazaro replied:
Just to be clear, is it ok to assume that the nodes leading up to this Fork will be processed only once by the backend and that their resulting PCollection instance will be used for all downstream nodes? So if there is re-computation it's entirely on Beam/Dataflow for re-computing a PCollection?

johnynek replied:
Yes. That's what the memoization is for. The whole `rec` bit is about caching.

tlazaro replied:
Cool, just wanted to confirm. I'm observing the following:

- Handwritten Beam code running on Dataflow does reuse PCollections and does a fan-out to multiple outputs. Not sure there are rules or hints can be passed but seems to just work. Will try to find out if there are limitations or if always reuses.
- The Scalding to Beam conversion for a real job is not doing reuse and is recomputing everything
  - One of our sources implementation doesn't have equality implemented and we though this was messing up memoization, we think we fixed it but still no reuse.
  - Trying even simpler cases, with TypedPipe.from(List(...)) and still can't get any reuse.
  
I will continue experimenting and narrowing it down, trying to understand if there is an issue with the conversion or in some assumption about Beam/Dataflow.

tlazaro replied:
I think we have found the problem, while we take care to end up with a single BeamOp instance per transform, the `run` in a BeamOp is called by every other transform that depends on it, causing recomputation. We could cache the result of the `run` locally in the BeamOp which seems easier or have a shared cache.

tlazaro replied:
Ok something like this does the trick:

```scala
  final case class TransformBeamOp[A, B](
    source: BeamOp[A],
    f: PTransform[PCollection[A], PCollection[B]],
    kryoCoder: KryoCoder,
    name: String
  ) extends BeamOp[B] {

    private lazy val memo = Memoize.function[Pipeline, PCollection[B]] {
      case (pipeline, _) =>
        val pCollection: PCollection[A] = widenPCollection(source.run(pipeline))
        pCollection.apply(name, f).setCoder(kryoCoder)
    }

    def run(pipeline: Pipeline): PCollection[B] = memo(pipeline)
  }
  ```

@johnynek do you think this approach makes sense? There are only a handful of BeamOp's so it isn't too much work. We could do the memo in the parent trait but the type of the elements wouldn't be as precise. We could cast or use some abstract types if feeling fancy, or just keep this duplication.

johnynek replied:
yeah, I would probably do something like below inside BeamOp:

```scala

protected def runNoCache(p: Pipeline): PCollection[_ <: A]
private val resultCache = scala.collection.mutable.Map.empty[Pipeline, PCollection[_ <: A]]
  
final def run(p: Pipeline): PCollection[_ <: A] =
  resultCache.getOrElseUpdate(p, runNoCache(p))
```

and then change all the current `def run` into `def runNoCache`

tlazaro replied:
This works by the way, got reuse on Dataflow.

johnynek replied:
Okay, good. Sorry my message missed yours. I would still do the `def runNoCache` on the base trait (but you are right, using `Memoize.function` is safer actually we found some bugs trying to do memoization with mutable Map in scala, so internally it uses a var + immutable map).

tlazaro commented:
We haven't tackled toIterableExecution and forceToDiskExecution yet. Still trying to get a successful run for a specific pretty large job.

tlazaro committed:
Minor fix

tlazaro committed:
Caching of BeamOp#run and CoGroup cleanup

tlazaro committed:
Adds shallow tests for BeamOp reuse

tlazaro commented:
Added some improvements to CoGroup so the TupleToKv doesn't show in the top level ui and rather lives inside CoGroup outer PTransform.

Added caching to BeamOp following @johnynek's advice, new implementations should by default be built using the cache preventing mistakes.

Added very shallow tests for the caching. Our goal should be testing the structure of the Pipeline in code, instead of visually in a Dataflow UI. The Pipeline DAG is a bit tedious to work with, the way would be to use the visitor pattern it has and perform mutations to build a better DAG instance where could perform better analysis.

tlazaro commented:
Can't afford to do proper structural testing now, though I'm sure will come back as a problem later.

johnynek commented on scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala:
```diff
@@ -35,6 +42,63 @@ class BeamBackendTests extends FunSuite with BeforeAndAfter {
+    val (pipeline, op) = beamPlan(a)
+
+    assert(op.isInstanceOf[FromIterable[Int]])
+    assert(op.run(pipeline) == op.run(pipeline))
```
do you want `eq` here? to test reference equality?

tlazaro replied:
yes my bad

tlazaro replied:
done

johnynek commented on scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala:
```diff
@@ -35,6 +42,63 @@ class BeamBackendTests extends FunSuite with BeforeAndAfter {
+    val (pipeline, op) = beamPlan(leftPipe.join(rightPipe))
+
+    assert(op.isInstanceOf[CoGroupedOp[Int, Int]])
+    assert(op.run(pipeline) == op.run(pipeline))
```
do you want `eq` here?

tlazaro replied:
done

johnynek commented on scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala:
```diff
@@ -35,6 +42,63 @@ class BeamBackendTests extends FunSuite with BeforeAndAfter {
+    val (pipeline, op) = beamPlan(a ++ b)
+
+    assert(op.isInstanceOf[MergedBeamOp[Int]])
+    assert(op.run(pipeline) == op.run(pipeline))
```
eq?

tlazaro replied:
done

johnynek commented on scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala:
```diff
@@ -35,6 +42,63 @@ class BeamBackendTests extends FunSuite with BeforeAndAfter {
+    val (pipeline, op) = beamPlan(source)
+
+    assert(op.isInstanceOf[BeamOp.Source[String]])
+    assert(op.run(pipeline) == op.run(pipeline))
```
`eq`?

tlazaro replied:
done

johnynek commented on scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala:
```diff
@@ -249,6 +352,28 @@ class BeamBackendTests extends FunSuite with BeforeAndAfter {
 
+  test("Merge (++) two pipes") {
+    val a = TypedPipe.from(Seq(5, 3, 2, 6, 1, 4))
+    val b = TypedPipe.from(Seq(15, 13, 12, 16, 11, 14))
```
maybe try with some duplicates to make sure `++` is uniquing the results?

tlazaro replied:
sure

johnynek commented on scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala:
```diff
@@ -249,6 +352,28 @@ class BeamBackendTests extends FunSuite with BeforeAndAfter {
+    )
+  }
+
+  test("Merge (++) many pipes") {
```
testing a `a ++ a` example is interesting since it should be optimized into a `flatMap` that just duplicates each element. Same with `a ++ a ++ a` I think.

tlazaro replied:
for sure

tlazaro replied:
I added a test to check merging a pipe to itself produces the right result but had a hard time trying to test the shape of the conversion, in a way that includes the optimizations. Do you think we can leave that work to a separate PR? I don't want to grow this one too much more.

tlazaro committed:
Fixes BeamBackend tests
